updates_and_signals/safe_message_handlers/src/cluster-manager.ts

```diff
@@ -0,0 +1,134 @@
+import * as wf from '@temporalio/workflow';
+import type * as activities from './activities';
+import { Mutex } from 'async-mutex';
+import {
+  AssignNodesToJobUpdateInput,
+  ClusterManagerState,
+  ClusterManagerStateSummary,
+  DeleteJobUpdateInput,
+} from './types';
+
+const { assignNodesToJob, unassignNodesForJob, startCluster } = wf.proxyActivities<typeof activities>({
+  startToCloseTimeout: '1 minute', // TODO
+});
+
+// ClusterManagerWorkflow keeps track of the job assignments of a cluster of nodes. It exposes an
+// API to started and shutdown the cluster, to assign jobs to nodes, to delete jobs, and to query
+// cluster status. The workflow maps this API to Signals, Updates, and Queries. The assign and
+// delete operations issue an RPC changing the state of the remote cluster, and then mutate workflow
+// state reflecting the change made. In order that workflow state remains in sync with the true
+// cluster state, assign/delete operations must not be performed concurrently (i.e. they must not
+// "interleave" with each other; they must be "serialized"; they must be "atomic"). An async mutex
+// from a 3rd party library is used to ensure this.
+export class ClusterManager {
+  state: ClusterManagerState;
+  jobsWithNodesAssigned: Set<string>;
+  nodesMutex: Mutex;
+
+  constructor(state?: ClusterManagerState) {
+    this.state = state ?? {
+      clusterStarted: false,
+      clusterShutdown: false,
+      nodes: new Map<string, string | null>(),
+      maxAssignedNodes: 0,
+    };
+    this.jobsWithNodesAssigned = new Set<string>();
+    this.nodesMutex = new Mutex();
+  }
+
+  async startCluster(): Promise<void> {
+    await startCluster();
+    this.state.clusterStarted = true;
+    for (let i = 0; i < 25; i++) {
+      this.state.nodes.set(i.toString(), null);
+    }
+    wf.log.info('Cluster started');
+  }
+
+  async shutDownCluster(): Promise<void> {
+    await wf.condition(() => this.state.clusterStarted);
+    this.state.clusterShutdown = true;
+    wf.log.info('Cluster shutdown');
+  }
+
+  async assignNodesToJob(input: AssignNodesToJobUpdateInput): Promise<ClusterManagerStateSummary> {
+    await wf.condition(() => this.state.clusterStarted);
+    if (this.state.clusterShutdown) {
+      // If you want the client to receive a failure, either add an update validator and throw the
+      // exception from there, or raise an ApplicationError. Other exceptions in the handler will
+      // cause the workflow to keep retrying and get it stuck.
+      throw new wf.ApplicationFailure('Cannot assign nodes to a job: Cluster is already shut down');
+    }
+    return await this.nodesMutex.runExclusive(async (): Promise<ClusterManagerStateSummary> => {
+      // Idempotency guard: do nothing if the job already has nodes assigned.
+      if (!new Set(this.state.nodes.values()).has(input.jobName)) {
```

Comment:
What if the job has already been assigned and completed? Do we care?

Reply:
Thanks. Switched to store seen job names in a set. This is what our Python version does. However it looks like our .NET sample has no idempotency guard cc @cretz.